src/convert_hl7v2_fhir/controllers/hl7/hl7_builder.py

```diff
@@ -0,0 +1,107 @@
+from datetime import datetime
+from enum import Enum, IntEnum
+from typing import Optional
+from uuid import uuid4, UUID
+
+
+class HL7ErrorCode(IntEnum):
+    """https://hl7-definition.caristix.com/v2/HL7v2.8/Tables/0357"""
+
+    ACCEPTED = 0
+    SEGMENT_SEQUENCE_ERROR = 100
+    REQUIRED_FIELD_MISSING = 101
+    DATA_TYPE_ERROR = 102
+    TABLE_VALUE_NOT_FOUND = 103
+    VALUE_TOO_LONG = 104
+    UNSUPPORTED_MESSAGE_TYPE = 200
+    UNSUPPORTED_EVENT_CODE = 201
+    UNSUPPORTED_PROCESSING_ID = 202
+    UNSUPPORTED_VERSION_ID = 203
+    UNKNOWN_KEY_IDENTIFIER = 204
+    DUPLICATE_KEY_IDENTIFIER = 205
+    APPLICATION_RECORD_LOCKED = 206
+    APPLICATION_INTERNAL_ERROR = 207
+
+
+class HL7ErrorSeverity(str, Enum):
+    """https://hl7-definition.caristix.com/v2/HL7v2.8/Tables/0516"""
+
+    ERROR = "E"
+    FATAL_ERROR = "F"
+    INFORMATION = "I"
+    WARNING = "W"
+
+
+def generate_ack_message(
+    receiving_application: str,
+    receiving_facility: str,
+    replying_to_msgid: str,
+    care_provider_email: Optional[str] = None,
+    care_provider_orgname: Optional[str] = None,
+    hl7_error_code: Optional[HL7ErrorCode] = None,
+    error_severity: Optional[str] = None,
+    error_message: Optional[str] = None,
+):
+    """https://hl7-definition.caristix.com/v2/HL7v2.8/TriggerEvents/ACK"""
+
+    message_header = _generate_msh_segment(receiving_application, receiving_facility)
+
+    segment_err = ""
+    segment_zha = ""
+    accept_code = ""
+
+    if hl7_error_code:
+        # generate an ACK based on rejecting the message and processing no further
+        accept_code = "AR"
+        # for ERR layout, see https://hl7-definition.caristix.com/v2/HL7v2.8/Segments/ERR
+        segment_err = f"\rERR|||{hl7_error_code}|{error_severity}||||{error_message}"
+    else:
+        # generate an ACK based on accepting the message and sending care provider email as response
+        accept_code = "AA"
+        if care_provider_email:
+            # custom defined segment (see tech docs)
+            segment_zha = f"\rZHA|{care_provider_orgname}|{care_provider_email}"
+
+    # see https://hl7-definition.caristix.com/v2/HL7v2.8/Segments/MSA
+    segment_msa = f"MSA|{accept_code}|{replying_to_msgid}"
+    return message_header + "\r" + segment_msa + segment_err + segment_zha
+
+
+def _generate_msh_segment(
+    receiving_application: str,
+    receiving_facility: str,
+    message_control_id: Optional[UUID] = None,
+):
+    """https://hl7-definition.caristix.com/v2/HL7v2.8/Segments/MSH"""
+
+    timestamp = datetime.now().strftime("%Y%m%d%H%M%S")

```

Comment:
I don't think this format is right..  We're not specifying a timezone, which according to the spec means it defaults to being interpreted by anything that speaks HL7v2 as the local timezone.  That means if we pass the output of this method to `to_fhir_datetime` below, which formats everything as UTC, we're actually an hour wrong for half the year.  If we specify `+0000` on the tail of the format string I *think* that removes any potential ambiguity.